File: main/hw1-heap-sort/hw1-heap-sort-byswap.java
import java.util.Scanner;

final class Solution {
    public static void main(final String[] args) {
        final int[] a = readRecord();
        Algo.heapsort(a);
        print(a);
    }

    private static int[] readRecord() {
        try (final Scanner sc = new Scanner(System.in)) {
            final int[] a = new int[sc.nextInt()];
            for (int i = 0; i != a.length; ++i) a[i] = sc.nextInt();
            return a;
        }
    }

    private static void print(final int[] a) {
        System.out.print("[");

        String sep = "";
        for (final int x : a) {
            System.out.format("%s%d", sep, x);
            sep = ",";
        }

        System.out.println("]");
    }

    private Solution() { throw new AssertionError(); }
}

final class Algo {
    static void heapsort(final int[] a) {
        int stop = a.length;
        if (stop < 2) return;

        // Convert the array into a binary maxheap.
        for (int parent = (stop - 1) / 2; parent != -1; --parent)
            sift_down(a, parent, stop);

        // Successively pop elements from the maxheap to sort the array.
        while (--stop != 0) {
            swap(a, 0, stop);
            sift_down(a, 0, stop);
        }
    }

    private static final int NO_CHILD = -1;

    private static void sift_down(final int[] a, int parent, final int stop) {
        for (; ; ) {
            final int child = pick_child(a, parent, stop);
            if (child == NO_CHILD || a[parent] >= a[child]) break;

            swap(a, parent, child);
            parent = child;
        }
    }

    private static int pick_child(final int[] a,
                                  final int parent, final int stop) {
        final int left = parent * 2 + 1;
        if (left >= stop) return NO_CHILD;

        final int right = left + 1;
        return right == stop || a[left] >= a[right] ? left : right;
    }

    private static void swap(final int[] a, final int i, final int j) {
        final int tmp = a[i];
        a[i] = a[j];
        a[j] = tmp;
    }

    private Algo() { throw new AssertionError(); }
}

File: main/hw1-heap-sort/hw1-heap-sort-byswap.py
#!/usr/bin/env python3


def read_value():
    """Reads a line of input as a single integer."""
    return int(input())


def read_record():
    """Reads a line of input as a whitespace-separated record of integers."""
    return map(int, input().split())


def read_input_values():
    """Reads input in the input format specified in the problem description."""
    length = read_value()

    values = []
    while len(values) < length:
        values.extend(read_record())

    if len(values) != length:
        raise ValueError('wrong input length')

    return values


def display_values(values):
    """Prints values in the output format given in the problem description."""
    print(f'[{",".join(map(str, values))}]')


def heapsort(values):
    """Heapsorts a list of values in nondecreasing order."""
    length = len(values)

    def pick_child(parent):
        left = parent * 2 + 1
        if left >= length:
            return None

        right = left + 1
        if right == length or values[left] >= values[right]:
            return left

        return right

    def sift_down(parent):
        while True:
            child = pick_child(parent)
            if child is None or values[parent] >= values[child]:
                break

            values[parent], values[child] = values[child], values[parent]
            parent = child

    # Convert the list into a maxheap.
    for parent in range(length // 2, -1, -1):
        sift_down(parent)

    # Extract each element from the maxheap, placing them from right to left.
    while length > 1:
        length -= 1
        values[0], values[length] = values[length], values[0]
        sift_down(0)


def run():
    """Reads the input values, heapsorts them, and displays the results."""
    values = read_input_values()
    heapsort(values)
    display_values(values)


if __name__ == '__main__':
    run()
